securedrop/crypto_util.py

```diff
@@ -12,51 +12,206 @@
+        u'P'
+
+        :param str name: The source's filesystem id (their codename, salted
+                         with SCRYPT_ID_PEPPER, and hashed with scrypt).
+        :param str secret: The source's codename.
+        :returns: a :class:`GenKey <gnupg._parser.GenKey>` object, on which
+                  the ``__str__()`` method may be called to return the
+                  generated key's fingeprint.
+
+        """
+        name = clean(name)
+        secret = self.hash_codename(secret, salt=self.scrypt_gpg_pepper)
+        return self.gpg.gen_key(self.gpg.gen_key_input(
+            key_type=self.GPG_KEY_TYPE,
+            key_length=self.__gpg_key_length,
+            passphrase=secret,
+            name_email=name
+        ))
+
+    def delete_reply_keypair(self, source_filesystem_id):
+        key = self.getkey(source_filesystem_id)
+        # If this source was never flagged for review, they won't have a reply
+        # keypair
+        if not key:
+            return
+        # The private key needs to be deleted before the public key can be
+        # deleted. http://pythonhosted.org/python-gnupg/#deleting-keys
+        self.gpg.delete_keys(key, True)  # private key
+        self.gpg.delete_keys(key)  # public key
+        # TODO: srm?
+
+    def getkey(self, name):
+        for key in self.gpg.list_keys():
+            for uid in key['uids']:
+                if name in uid:
+                    return key['fingerprint']
+        return None
+
+    def encrypt(self, plaintext, fingerprints, output=None):
+        # Verify the output path
+        if output:
+            current_app.storage.verify(output)
+
+        if not isinstance(fingerprints, (list, tuple)):
+            fingerprints = [fingerprints, ]
+        # Remove any spaces from provided fingerprints GPG outputs fingerprints
+        # with spaces for readability, but requires the spaces to be removed
+        # when using fingerprints to specify recipients.
+        fingerprints = [fpr.replace(' ', '') for fpr in fingerprints]
+
+        if not _is_stream(plaintext):
+            plaintext = _make_binary_stream(plaintext, "utf_8")
+
+        out = self.gpg.encrypt(plaintext,
+                               *fingerprints,
+                               output=output,
+                               always_trust=True,
+                               armor=False)
+        if out.ok:
+            return out.data
+        else:
+            raise CryptoException(out.stderr)
 
-# map code for a given language to a localized wordlist
-language2words = {}  # type: Dict[Text,List[str]]
-nouns = open(config.NOUNS).read().rstrip('\n').split('\n')
-adjectives = open(config.ADJECTIVES).read().rstrip('\n').split('\n')
-
-
-class CryptoException(Exception):
-    pass
+    def decrypt(self, secret, ciphertext):
+        """
+        >>> key = self.genkeypair('randomid', 'randomid')
+        >>> self.decrypt('randomid',
+        ...   self.encrypt('Goodbye, cruel world!', str(key))
```

Comment:
This string was changed in #2982, let's keep those changes

Reply:
Fixed